applications/FluidDynamicsApplication/custom_processes/shock_detection_process.h

```diff
@@ -0,0 +1,366 @@
+        if (r_comm.IsDistributed()) {
+            for (auto &r_node : r_comm.LocalMesh().Nodes()) {
+                auto& r_gp_to_neighbours = r_node.GetValue(NEIGHBOUR_NODES).GetContainer();
+                for (auto &r_gp : r_gp_to_neighbours) {
+                    global_pointers_list.push_back(r_gp);
+                }
+            }
+            global_pointers_list.Unique();
+        }
+
+        // Now create the pointer communicator and shock values retrieve proxy
+        GlobalPointerCommunicator<Node<3>> pointer_communicator(r_data_comm, global_pointers_list);
+        auto shock_variables_proxy = pointer_communicator.Apply([&](const GlobalPointer<Node<3>>& rpNode)
+        {
+            NeighbourData<TShockVariableType, TShockGradientVariableType> neighbour_data(
+                rpNode->FastGetSolutionStepValue(rShockVariable),
+                rpNode->GetValue(rShockGradientVariable),
+                rpNode->Coordinates());
+            return neighbour_data;
+        });
+
+        // Perform the shock detection
+#pragma omp parallel for
+        for (int i_node = 0; i_node < static_cast<int>(r_comm.LocalMesh().NumberOfNodes()); ++i_node) {
+            auto it_node = r_comm.LocalMesh().NodesBegin() + i_node;
+            double& r_shock_sens = it_node->GetValue(mrShockSensorVariable);
+            const auto& r_var_i = it_node->FastGetSolutionStepValue(rShockVariable);
+            const auto& r_grad_var_i = it_node->GetValue(rShockGradientVariable);
+
+            // Loop the neighbours to compute the shock sensor
+            r_shock_sens = 0.0;
+            const double zero_tol = 1.0e-8;
+            auto& r_neighbours = it_node->GetValue(NEIGHBOUR_NODES);
+            KRATOS_DEBUG_ERROR_IF(r_neighbours.size() == 0) << "Node " << i_node << " has no neighbours." << std::endl;
+            for (auto& r_neigh : r_neighbours ) {
+                // Get the neighbour values
+                const auto values_j = shock_variables_proxy.Get(&r_neigh);
+                const double& r_var_j = values_j.mShockVariableValue;
+                const auto& r_grad_var_j = values_j.mShockGradientVariableValue;
+                const auto l_ji = values_j.mCoordinates - it_node->Coordinates();
+
+                // Calculate the shock sensor auxiliary values
+                const auto aux_1 = r_var_j - r_var_i;
+                const auto aux_2 = 0.5 * inner_prod(l_ji, r_grad_var_i + r_grad_var_j);
+                const auto num = aux_1 - aux_2;
+                const auto den = std::abs(aux_1) + std::abs(aux_2);
+                double beta_ij = 0.0;
+
+                // Check if the solution is not constant (den close to 0.0)
+                if (std::abs(den) > zero_tol) {
+                    // Compute and bound the shock sensor
+                    const double aux_beta_ij = std::abs(num / den);
+                    beta_ij = aux_beta_ij < 1.0 ? aux_beta_ij : 1.0;
+                } else {
+                    beta_ij = 0.0;
```

Comment:
the else is not needed since you initialize beta to zero

Reply:
Done. Thanks for spotting it!